sql/table.h

```diff
@@ -1870,6 +1870,14 @@ enum enum_schema_table_state
 enum enum_fk_option { FK_OPTION_UNDEF, FK_OPTION_RESTRICT, FK_OPTION_NO_ACTION,
   FK_OPTION_CASCADE, FK_OPTION_SET_NULL, FK_OPTION_SET_DEFAULT };
 
+enum enum_fk_nullablity
+{
+  /** Foreign table or Referenced table doesn't exist */
+  NOT_EXIST,
+  ALLOWS_NULL,
+  ALLOWS_NOT_NULL
```

Comment:
The literals `ALLOWS_NULL` and `ALLOWS_NOT_NULL` need to be documented. In particular, I want to know what would be the values when there is a `CONSTRAINT FOREIGN KEY (a_null,b_not_null) REFERENCES parent(c_not_null,d_null)`.

The logic of `get_foreign_key_info()` looks like `ALLOWS_NULL` will take priority, that is, we would only set `ALLOWS_NOT_NULL` if a suitable index is found and `ALLOWS_NULL` was not set yet. It would be more descriptive to use the name `INDEX_EXISTS` instead of `ALLOWS_NOT_NULL`.

Reply:
I thought more about this, and I think that we need to match each column. This might require ‘pushing down’ the check to InnoDB.
Consider the valid case `CONSTRAINT FOREIGN KEY (a_null,b_not_null) REFERENCES parent(c_null,d_not_null) ON UPDATE CASCADE`. As far as I understand, with the current logic this would be disallowed due to `ALLOWS_NULL` being reported for `parent.c_null`. But, there is no actual problem because `a_null` does allow `NULL` values.

I think that the test case must be expanded with 2-column `FOREIGN KEY` constraints like this, covering all 4 combinations of `NULL` and `NOT NULL` for the 2+2 columns.

Reply:
I think that @FooBarrior would be a good candidate for reviewing the changes outside InnoDB, related to this.